### PR TITLE
Harvest Getty by record id instead of offset paging

### PR DESCRIPTION
**Pull request by DominicBM**
Getty has been on hold since its last good harvest in **February 2021**. This gets it running again: **101,384 of 101,393 records (99.99%)**, mapping 101,384 attempted / 101,384 successful / **0 failed**, validated against the live endpoint.

## The problem

Getty's Ex Libris Primo gateway caps any single query at `offset <= 1999` and `limit <= 1000`. `GettyHarvester` pages by offset, so it retrieves ~2,000 of ~101,400 records — **and reports success**. That silent 98% shortfall reached production in February 2026.

It is not rate limiting. No retry, backoff, or page-size change moves it.

## The approach

A single record *is* retrievable by id, with no offset involved:

```
q=rid,exact,GETTY_ROSETTAIE10176553   ->   exactly 1 record
```

`GettyRefreshHarvester` looks up every known record id directly, then asks Getty's own `newrecords` facet for anything added in the last 90 days.

The seed is `getty.harvest.seed` — the previous harvest's `OriginalRecord.avro`. Each run's output becomes the next run's seed, so discovered ids carry forward automatically instead of by hand.

## Hardening

This makes ~101K requests against a partner's production system, so:

- **Route-loss breaker** — aborts after 3 consecutive 403s rather than working through the whole seed against an endpoint that has already refused us.
- **Fail-fast route probe** — the first lookup verifies the allowlisted egress is applied.
- **Resolved-fraction floor** — refuses to publish if under half the seeded ids resolve, so a collapsed harvest cannot quietly replace a good one.
- **Serialised Avro writes** — the writer inherited from `LocalHarvester` is not thread-safe.
- **Pacing after the body is read** — sleeping inside the request context holds an idle connection open and the server drops it. That looks like rate limiting and gets "fixed" by raising the delay, which makes it strictly worse. This one cost hours to diagnose; there is a comment on it.

23 new unit tests (1,261 total, all passing). Everything load-bearing — URL construction, id extraction, the route-loss signal, seed parsing — lives in the companion object so it tests without Spark or a network.

## This is temporary, and it does not guarantee completeness

Stated plainly in the scaladoc, `SCRIPTS.md`, `README_INGESTS.md`, i3.conf and Confluence:

| Gap | Why | Consequence |
|---|---|---|
| 90-day discovery window | `newrecords` offers only 07/30/90 days back, cumulative; 90 is the longest | A quarterly schedule has **no margin**. A run that slips leaves records invisible to discovery. |
| `GETTY_OCP` cannot be enumerated | Measured 2026-09-12: excluding the ROSETTA subset, the remaining 78,613 records collapse to one `rtype` value and one `local2` value. Nothing to partition on. | OCP coverage (77.5% of the collection) rests **entirely on the seed**. It has held only because OCP has not gained a record since 2021 — its max id is unchanged. |
| 9 records short | Facet reports 101,393; harvest holds 101,384 | Under 0.01%, all on the ROSETTA side. |

It keeps the aggregation from drifting. It is not a guarantee that DPLA holds every record Getty publishes. The real fix is Ex Libris lifting the cap or Getty providing a bulk feed — which is worth continuing to press for.

**`GettyHarvester` is retained but no longer wired up.** It becomes correct again the moment the cap is lifted; `GettyProfile` just needs re-pointing.

## Also here

- `getty.harvest.seed` added to the conf schema.
- A note on `PrimoVEHarvester.scala:74`: it extracts ids as `(doc \\ "control" \ "recordid").toString`, which yields the json4s AST's `toString` rather than the id (there is already a `FIXME` on the line). Harmless while ids are only written out, but it would corrupt exactly the seed round-trip this design depends on — so the new harvester extracts cleanly and a test pins it. **This also affects MWDL and Mississippi and is worth a separate look.**

## Before first run

`getty.harvest.seed` points at `s3://dpla-master-dataset/getty/harvest/20260912_213156-getty-OriginalRecord.avro`, which currently exists **only on the ingest EC2** — it has not been synced to S3. Either run `./scripts/s3-sync.sh getty harvest` or point the seed at the local path. Flagged in i3.conf too.

Config changes are already on `ingestion3-conf` master (`7cae868`, `cff6f0b`, and follow-ups): Getty off hold, quarterly Mar/Jun/Sep/Dec.

🤖 Generated with [Claude Code](https://claude.com/claude-code)

## Summary

- Replaces Getty offset paging with direct record-ID lookups.
- Discovers new records through Getty’s 90-day `newrecords` facet.
- Selects the newest completed harvest as the seed. Supports `getty.harvest.seed` for backfills.
- Adds route checks, request pacing, serialized Avro writes, failure propagation, and resolved-record guardrails.
- Maps valid rights URIs to `edmRights` and deduplicates thumbnail links.
- Updates Getty configuration and documents the bi-monthly schedule and discovery limits.
- Adds 38 Getty harvester tests and script tests. Repository validation reports 1,276 passing tests.
- Formats partner email recipient lists for readability.
- Logs a `GETTY DISCOVERY GAP` warning after more than 90 days without a completed harvest.

## Operational impact

- A manual pipeline trigger or ECS redeployment is not specified.
- No environment variables or AWS Secrets Manager keys are changed.
- No database migration is included.
- No shared infrastructure configuration changes are included.
- No public API response shapes or endpoints are changed.
- The harvester uses existing Getty credentials and adds direct external record lookups and route checks.

### CODE REVIEW

**Comment by DominicBM** [comment]
**Update — the 90-day window is now mitigated on both sides.**

Two changes since the PR was opened, in `d9ac897e`:

**1. Getty moves to bi-monthly, not quarterly** (`ingestion3-conf` master). Discovery reaches back only 90 days, so a quarterly cadence sits right on the edge — a run that slips by a few days opens a permanent blind spot. Jan/Mar/May/Jul/Sep/Nov leaves about a month of margin.

**2. The harvester warns when the interval exceeds the window anyway.** `GettyRefreshHarvester` reads the previous harvest's date out of the seed's activity path and, if more than 90 days have passed, logs a banner:

```
GETTY DISCOVERY GAP: 181 days have passed since the previous harvest (2026-09-12), but
Getty's `newrecords` facet only reaches back 90 days. Records added between 2026-09-12
and 2026-12-12 are in neither the seed nor the discovery window, and no later run will
find them. This harvest cannot be treated as complete.
```

It names the exact date range that was lost, so the operator knows what is unrecoverable rather than just that something is wrong.

It **warns rather than fails** on purpose: a late harvest is still far better than none, and refusing to run would widen the very gap being reported.

10 more tests (33 in this suite, 1,271 total, all passing), including the edge case that 90 days exactly should stay quiet — crying wolf on a schedule that is working would train people to ignore it — and that a text-file seed carrying no date says nothing rather than guessing.

Docs updated to match: `SCRIPTS.md`, `README_INGESTS.md`, i3.conf, and both Confluence pages.

**Comment by DominicBM** [comment]
**Update — the seed is now discovered, not configured** (`7d5dbc1d`).

Pinning a specific harvest directory in `getty.harvest.seed` was wrong: it would have meant editing `i3.conf` after every ingest, and that is run state, not configuration. It is also state the code can derive, since `OutputHelper` writes activity directories as `YYYYMMDD_HHMMSS-<hub>-<schema>` — a convention this codebase owns.

`GettyRefreshHarvester` now lists the hub's harvest activity directory and takes the newest completed run. The timestamp is fixed-width and zero-padded, so lexical order is chronological order and no date parsing is needed to rank them.

Two things it is careful about:

- **Skips directories without `_SUCCESS`.** A crashed Spark write leaves a `_temporary` directory behind; seeding from one would silently shrink the next harvest.
- **Exact hub match.** `getty` must not match `getty-test` — seeding from the wrong hub would be silent and catastrophic. There is a test for it.

The output root is a run parameter (`--output`), not configuration, so it reaches the harvester through a new `SeedsFromPreviousHarvest` mixin that `HarvestExecutor` populates. Harvesters that do not mix it in are untouched — one `match` in the executor, no signature changes.

`getty.harvest.seed` stays in the schema as an override for backfills and re-seeding, but is no longer set in `i3.conf`, so **routine ingests now need no config change at all**. That also removes the "before first run" caveat from the PR description: there is no longer a hardcoded S3 path that needs a sync to exist first — the harvester will pick up whatever the newest completed harvest is, local or S3.

Five more tests (38 in this suite, 1,276 total, all passing).

File: src/main/scala/dpla/ingestion3/harvesters/api/GettyRefreshHarvester.scala [hunk 1]
```diff
@@ -0,0 +1,628 @@
...
+      if (!fs.exists(path)) Seq.empty
+      else
+        fs.listStatus(path)
+          .filter(_.isDirectory)
+          .map(_.getPath.getName)
+          .toSeq
+          // Skip incomplete Spark writes: a directory with _temporary and no
+          // _SUCCESS is a crashed run, and seeding from one would silently
+          // shrink the harvest.
+          .filter(name => fs.exists(new HPath(s"$dir/$name/_SUCCESS")))
```
[the remark this answers is not in the document]
Comment (DominicBM) [reply src/main/scala/dpla/ingestion3/harvesters/api/GettyRefreshHarvester.scala:196]: Declining this one, with thanks for the precise trace — the premise is correct but I don't think the conclusion follows.

You're right that `HarvestExecutor` writes `_SUCCESS` during `save(outputPath)` and only logs a warning if the later `writeManifest(...)` returns `Failure`, so `_SUCCESS` without `_MANIFEST` is reachable.

But `_SUCCESS` is Spark's commit marker: it is written only after every partition has landed. Its presence already establishes that the Avro is complete, which is the property the seed selection actually depends on. A missing `_MANIFEST` means DPLA metadata didn't get written afterward — not that the records are partial.

So requiring both would:

- **not** increase confidence in the data being seeded from, since `_SUCCESS` alone already guarantees it, and
- risk rejecting a sound harvest because a metadata file failed to write (an S3 hiccup, say), forcing a manual override at exactly the wrong moment — the start of a multi-hour run.

It also wouldn't catch the failure worth worrying about here. A harvester that returns a genuinely short DataFrame gets *both* markers written; requiring `_MANIFEST` adds nothing against it. That case is guarded separately, by `MinResolvedFraction` and — as of 85522bcb — by aborting when any seeded lookup ends in `LookupFailed`.

Leaving the filter on `_SUCCESS` alone. Happy to revisit if a harvest ever turns up where `_SUCCESS` is present and the data isn't complete, since that would undermine the assumption above rather than just this check.

**Comment by DominicBM** [comment]
@coderabbitai full review

**Comment by DominicBM** [comment]
@coderabbit full review

**Comment by DominicBM** [comment]
## Harvest-side follow-up: `Gone` is not retried, and a transient empty response is indistinguishable from a deletion

Recording this here so it is preserved with the code it concerns. **Not a blocker for this PR** — it is a design consequence that surfaced while designing the decoupled `newrecords` discovery job, and it wants fixing in a follow-up.

### What is already correct — don't "fix" this

`LookupFailed` is handled properly. [Line 320](https://github.com/dpla/ingestion3/blob/feature/getty-id-refresh-harvester/src/main/scala/dpla/ingestion3/harvesters/api/GettyRefreshHarvester.scala#L320) refuses to publish if any seeded lookup ended indeterminate, precisely so an id cannot vanish from the next seed because of a transient error. That guard is right and should stay.

### The hole

[Line 374](https://github.com/dpla/ingestion3/blob/feature/getty-id-refresh-harvester/src/main/scala/dpla/ingestion3/harvesters/api/GettyRefreshHarvester.scala#L374):

```scala
case Gone => return last     // no retry; only LookupFailed loops
```

`Gone` means the request **succeeded** and carried zero documents. It is never retried. And we have reproduced Primo emitting **HTTP 200 with `docs: []` transiently** — see #777, comment of 2026-09-13, §2: intermittent rather than positional, so it cannot be avoided by knowing which requests are risky.

The chain:

1. transient `200` + `docs: []`
2. parses cleanly → `docsOf(json).headOption` is `None` → `Gone`
3. returned immediately, no retry
4. no record written
5. id absent from the harvest output → **absent from the next seed** → never asked about again

**One blip, one record lost, permanently.** Silent, because `gone > 0` is the expected case — nothing distinguishes five transient blips from five genuine withdrawals. The only backstop is `MinResolvedFraction = 0.5`.

### Why this is not hypothetical

Mississippi's corpus measured **~138,800 on 11–13 September and ~117,300 on 18 September** — two independent methods (`any,contains,**` and an a–z `begins_with` sweep, neither using wildcards) agreeing on a ~15% drop. Had a refresh run that day against a 138,565-id seed, roughly **21,000 ids would have resolved `Gone`** and dropped out of the seed. At 85% resolved, the 0.5 floor does not fire.

Whether those records were genuinely withdrawn or the view was mid-reindex, the harvester cannot tell — and under the current design the ids are gone either way.

### Proposed fix (follow-up PR)

1. **Retry `Gone` in-run** — two attempts with backoff before believing it. Kills most blips at the source, ~3 lines, independent of everything below.
2. **Seed from a cumulative id store, not from the previous harvest's output.** The decoupled discovery job is already building an append-only id store; make that the authoritative seed. An id that blips `Gone` is then re-asked next run and self-heals, instead of being dropped forever.
3. **Tombstones to bound it.** Retire an id only after it returns `Gone` on **N consecutive harvests** (3 is plenty). Transients heal; real deletions cost three cycles of one request each, then stop. Discovery stays dumb and additive; the harvest owns the tombstone file.
4. **Compute `MinResolvedFraction` against live ids** (cumulative minus confirmed tombstones), or it drifts downward as unretired dead ids accumulate and eventually trips falsely.
5. **Add a deletion-rate circuit breaker.** The 0.5 floor catches catastrophe, not the 15% case above. Abort when the share of previously-resolving ids going `Gone` exceeds a few percent, and make a human look.

### Note on what discovery can and cannot check

Discovery cannot reconcile its id count against the corpus total: that total is the **net** of additions and deletions, so a hub adding 500 and removing 500 looks unchanged. Corpus size is a **tripwire** worth logging — a −21,500 swing should wake someone — but it is not a completeness check, and shouldn't be built as one.
